app/api/puzzle/nextPuzzle/nextFor.ts

```
@@ -0,0 +1,133 @@
+    return undefined;
+  }
+  return puzzleFromDocument(p);
+};
+
+const nextPuzzleRepetitions = async (
+  rating: number,
+  reps: number,
+  ratingMap: Map<string, Rating>,
+  exceptions: any
+): Promise<Puzzle> => {
+  if (reps == MAX_REPS) {
+    throw new Error('Maximum repetitions reached... something is wrong.');
+  }
+  const original = frequentiallyRandomTheme();
+  const mappedRating = ratingMap.get(original);
+  // If the theme is not in the rating map, let's try to use it.
+  if (!mappedRating) {
+    const p = await nextPuzzleForThemeAndRating(original, rating, exceptions);
+    if (p) {
+      console.log(`Found theme ${original} frequentially after ${reps} reps.`);
+      return p;
+    }
+    return await nextPuzzleRepetitions(rating, reps + 1, ratingMap, exceptions);
+  }
+  // Otheriwse, let's run probabilistic SM2 on the rating map.
+  const theme = sm2RandomThemeFromRatingMap(ratingMap);
+  const p = await nextPuzzleForThemeAndRating(theme, rating, exceptions);
+  if (p) {
+    console.log(`Found theme ${theme} through SM2 after ${reps} reps .`);
+    return p;
+  }
+  return await nextPuzzleRepetitions(rating, reps + 1, ratingMap, exceptions);
+};
+
+const nextPuzzleFor = async (user: User): Promise<PuzzleWithUserRating> =>
+  fetchUserRating(user).then(async ({ userRating, present }) => {
+    if (!present) {
+      RatingColl.create({
+        username: user.username,
+        rating: userRating.rating,
+        ratingDeviation: userRating.ratingDeviation,
+        volatility: userRating.volatility,
+        numberOfResults: userRating.numberOfResults,
+      });
+      // NB: This is important. If new user, set solved to be empty.
+      AllRoundColl.create({
+        username: user.username,
+        solved: [],
+      });
+    }
+
+    const ratingMap = await getThemeRatings(user);
+
+    // TODO: Better handle repeat avoidance.
+    const exceptions = (await AllRoundColl.findOne({ username: user.username }))
+      ?.solved ?? [];
```

The current implementation of submit requires an empty allRound entry for the user (hence [this comment](https://github.com/ahmedmahmud/lychee/blob/2b246a6446bca58cc2a9fb5b15bcccdee3599b8f/app/api/puzzle/nextPuzzle/nextFor.ts#L87)) so I'd actually prefer to fail loudly here.